### PR TITLE
Support @Blocking annotation for gRPC services

### LINKED ISSUE #4389: Support `@Blocking` annotation for gRPC services
Users can only specify to run a service on a blocking executor using `GrpcServiceBuilder.useBlockingTaskExecutor(true)`.
https://github.com/line/armeria/blob/603e2fb23b7227c4117b5b8d7377314152f9feb4/grpc/src/main/java/com/linecorp/armeria/server/grpc/GrpcServiceBuilder.java#L678
If `@Blocking` annotation is supported for gRPC services or methods:
https://github.com/line/armeria/blob/8bfe0be035ad0f2d01a6ac5ebeae465ee0a552e1/core/src/main/java/com/linecorp/armeria/server/annotation/Blocking.java#L33
- Users can easily know whether a service runs on a blocking executor.
- Selectively apply a blocking executor to the methods.
```java
class MyGrpcServiceImpl extends MyGrpcServiceImplBase {
    @Blocking
    public void blockingCall(...) {
        ...
    }

    public void nonBlockingCall(...) {
        ...
    }
}

@Blocking
class MyBlockingGrpcServiceImpl extends MyGrpcServiceImplBase {
    
    public void blockingCall1(...) {
        ...
    }

    public void blockingCall2(...) {
        ...
    }
}
```

### PR DESCRIPTION
**Pull request by ta7uw**
Motivation:

It is described in the following Issue.
https://github.com/line/armeria/issues/4389

Modifications:

- When building `HandlerRegistry`, find `ServerMethodDefinition`  that needs to be executed on the blocking executor
- In `FramedGrpcService#startCall`, in addition to `useBlockingTaskExecutor` field, check if the blocking executor is used for each `ServerMethodDefinition`.

Result:

- Closes #4389
- Users can easily know whether a service runs on a blocking executor.
- Selectively apply a blocking executor to the methods.

```java
class MyGrpcServiceImpl extends MyGrpcServiceImplBase {
    @Blocking
    public void blockingCall(...) {
        ...
    }

    public void nonBlockingCall(...) {
        ...
    }
}

@Blocking
class MyBlockingGrpcServiceImpl extends MyGrpcServiceImplBase {
    
    public void blockingCall1(...) {
        ...
    }

    public void blockingCall2(...) {
        ...
    }
}

```

### CODE REVIEW

File: grpc/src/main/java/com/linecorp/armeria/server/grpc/HandlerRegistry.java [hunk 1]
```diff
@@ -93,20 +97,24 @@ final class HandlerRegistry {
     private final Map<ServerMethodDefinition<?, ?>, List<DecoratorAndOrder>> annotationDecorators;
     private final Map<ServerMethodDefinition<?, ?>, List<? extends Function<? super HttpService,
             ? extends HttpService>>> additionalDecorators;
+    private final Set<ServerMethodDefinition<?, ?>> needToUseBlockingMethodDefs;
```
Comment (minwoox) [inline grpc/src/main/java/com/linecorp/armeria/server/grpc/HandlerRegistry.java:100]: nit: How about just `blockingMethod`?

**Review by minwoox — APPROVED** [review]
Just one nit.
Great job, @ta7uw! 😄

**Review by minwoox — COMMENTED** [review]
Could you check the test failure?
https://github.com/line/armeria/actions/runs/3762899880/jobs/6395997150#step:7:1321

**Comment by ta7uw** [comment]
@minwoox 
How can I check the test result report on CI?

**Comment by ikhoon** [comment]
> @minwoox How can I check the test result report on CI?

I believe you can download `reports-JVM-xx` from the `Artifacts` section on the GH actions summary.
https://github.com/line/armeria/actions/runs/3774443442

**Comment by Dogacel** [comment]
Should we add another annotation, `@Nonblocking` just for consistency?

So we can define the default behavior in a class to be `Blocking` but individual methods can run in a non-blocking fashion.

**Comment by minwoox** [comment]
Hi, thanks for the idea. 😄 
Well, that's probably useful when a service has a bunch of blocking methods and a few non-blocking methods.
However, because everything is non-blocking by default in Armeria, I'm not sure if it's a good idea to add the annotation.
Anyway, let me bring this up with the team and discuss it. Thanks!

**Comment by ikhoon** [comment]
I prefer adding a boolean parameter to `@Blocking` annotation rather than introducing a new annotation. Users may use `@Blocking(false)` if they want to execute a service or a method in an event loop.
```java
public @interface Blocking {
    boolean value() default true;
}
```

**Review by jrhee17 — COMMENTED** [review]
Looks good!

My guess is that `blockingTaskExecutor` is used in a lot of places so it is difficult to assume the usage is limited to executing `@Blocking` annotated methods.
(This can be seen just by querying usages of `ServiceRequestContext#blockingTaskExecutor`...)

Apart from the flaky tests, looks good to me 👍

**Comment by minwoox** [comment]
No, we don't have to add it in this PR. 😄

**Review by ikhoon — APPROVED** [review]
Thanks, @ta7uw. You are the gRPC hero! 😆

**Review by jrhee17 — APPROVED** [review]
Looks very clean! Thanks @ta7uw 🙇 🚀 👍

**Comment by ta7uw** [comment]
Would you like me to create an issue about adding a boolean parameter to `@Blocking` annotation ?

**Comment by minwoox** [comment]
It's up to you. 😄 
We can create an issue or directly make the PR.

**Comment by ta7uw** [comment]
OK.
I will create the issue.

**Pull request merged**